### DISCUSSION TITLE
ItemsControl.ItemsControlFromItemContainer issue with ListView on android

### DISCUSSION THREAD

**Discussion by gfortin13** [unoplatform/uno#9077] [General]
Hey team 👋 Guillaume Fortin (ex Nventive employee) here 😄 

For context I need to access my `ListView`'s `SelectedIndex` from inside a `DataTemplateSelector` for some implementation purpose.

I learned about the `ItemsControl.ItemsControlFromItemContainer(container)` method and by simply casting `(ListView)` to the result, since it inherits from `ItemsControl`, I am able to achieve this in UWP.

On android, `ItemsControl.ItemsControlFromItemContainer(container)` returns null. Is this a bug on the Uno side? Expected behavior? In all cases, in the meantime is there an alternative to finding my `ListView` from inside my DataTemplateSelector?

> Note, `ItemsControl.ItemsControlFromItemContainer(container)` works on android if the xaml implements a `ItemsControl`, just not a `ListView`, but I need the selected index.

**Comment by jeromelaban** [comment]
Thanks for the report Guillaume! Glad to see you here ;) [`ItemsControlFromItemContainer` is implemented](https://github.com/unoplatform/uno/blob/1d0745f8e3af6172e437e401cdc4e7e7b0f04082/src/Uno.UI/UI/Xaml/Controls/ItemsControl/ItemsControl.cs#L1538), but it may not behave the right way.

Cam you create a repro sample to show the incorrect behavior?

**Reply by gfortin13** [reply]
Thanks for the quick response Jerome! Might be small enough to just paste here, the small simple data template selector I use:

```
	public class RangeItemTemplateSelector : DataTemplateSelector
	{
		public DataTemplate UnselectedItems { get; set; }

		public DataTemplate SelectedItems { get; set; }

		protected override DataTemplate SelectTemplateCore(object item, DependencyObject container)
		{
			var itemsControl = ItemsControl.ItemsControlFromItemContainer(container);
			var currentIndex = itemsControl.IndexFromContainer(container);

			if (currentIndex > ((ListView)itemsControl).SelectedIndex)
			{
				return UnselectedItems;
			}
			else
			{
				return SelectedItems;
			}
		}
	}
```

Used in a list view

```
<ListView ItemsSource="{Binding ArbitrarySource}"
		SelectedIndex="{Binding ArbitraryIndex, Mode=TwoWay}"
		SelectionMode="Single"
		ItemTemplateSelector="{StaticResource RangeItemTemplateSelector}">
	<ListView .ItemsPanel>
		<ItemsPanelTemplate>
			<StackPanel Orientation="Horizontal" />
		</ItemsPanelTemplate>
	</ListView .ItemsPanel>
</ListView >
```

For context the template selector implementation mimics what the RatingControl kinda does, by having a selected state for all items before and including the selected index, and I can have a dynamic amount of stars (or any other range selectable items).

Not sure if it matters that I'm still on Uno.UI 3.1.6, the implementation is there and seems identical. On UWP this works, on android `itemsControl ` is null. If I use an `ItemsControl` instead, android can find it, but then I lose out on `.SelectedIndex`.

In the meantime, any other way to find my `ItemsControl`/`ListView` from the template selector?

**Reply by jeromelaban** [reply]
It looks like the property is not yet set for the container when it is provided to the Template Selector. It should be relatively easy to fix, though, to ensure that it is set at the right time. Fixing it for Uno 3.1.x is unlikely to happen, though.

**Reply by jeromelaban** [reply]
Let's track it in https://github.com/unoplatform/uno/issues/9081

**Reply by gfortin13** [reply]
Any other way to find the ListView my item belongs to inside the template selector? Even if it's hacky in the meantime.

**Reply by jeromelaban** [reply]
You could try an element name binding directly on the `DataTemplateSelector` entry, but it may not work:
```csharp
public class RangeItemTemplateSelector : DataTemplateSelector
{
   ItemsControl Parent { get; set; }
}
```
Then use it like this:
```xaml
<ListView x:Name="myListView" 
		ItemsSource="{Binding ArbitrarySource}"
		SelectedIndex="{Binding ArbitraryIndex, Mode=TwoWay}"
		SelectionMode="Single">
	<ListView.ItemTemplateSelector>
	   <local:RangeItemTemplateSelector Parent="{Binding ElementName=myListView}" />
	</ListView.ItemTemplateSelector>
	<ListView .ItemsPanel>
		<ItemsPanelTemplate>
			<StackPanel Orientation="Horizontal" />
		</ItemsPanelTemplate>
	</ListView .ItemsPanel>
</ListView >
```

**Comment by gfortin13** [comment]
Hi @jeromelaban , I restarted working on my project recently and updated to the most recent UnoApplicationTemplate (which has Uno version 4.4.20).

I noticed this issue is still happening on Android, any idea if this was going to be addressed at some point? I have been trying to use this functionality to make some sort of a "RangeSelector" as I've been calling it that looks like this

![image](https://github.com/unoplatform/uno/assets/3281015/ca57c82b-bd27-457e-868b-2560ad13bcca)
or like this
![image](https://github.com/unoplatform/uno/assets/3281015/04703258-69a6-4244-8dc4-73d1c1217ae3)

Using the current index of the list item vs the selected index in the template selector to determine if the item should be "full" or "empty" based on the range I am selecting.

If there is no short future in sight for this issue to be resolved, would you be able to provide an alternative
